NPE throw when get bucket count from MetadataUpdater

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.


### Fluss version

0.8.0 (latest release)

### Please describe the bug 🐞

The error is as follow:

```
2025-11-30 13:21:54
java.io.IOException: Failed to send record to table fluss.dwd_xxx. Writer state: running
	at org.apache.fluss.flink.sink.writer.FlinkSinkWriter.write(FlinkSinkWriter.java:159)
	at org.apache.flink.streaming.runtime.operators.sink.SinkWriterOperator.processElement(SinkWriterOperator.java:182)
	at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask$StreamTaskNetworkOutput.emitRecord(OneInputStreamTask.java:245)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.processElement(AbstractStreamTaskNetworkInput.java:217)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.emitNext(AbstractStreamTaskNetworkInput.java:169)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:68)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:616)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:231)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:1071)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:1020)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:968)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:947)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:754)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:569)
	at java.lang.Thread.run(Thread.java:879)
Caused by: org.apache.fluss.exception.FlussRuntimeException: Failed to send record to table fluss.dwd_xxx. Writer state: running
	at org.apache.fluss.client.write.WriterClient.doSend(WriterClient.java:216)
	at org.apache.fluss.client.write.WriterClient.send(WriterClient.java:139)
	at org.apache.fluss.client.table.writer.AbstractTableWriter.send(AbstractTableWriter.java:66)
	at org.apache.fluss.client.table.writer.AppendWriterImpl.append(AppendWriterImpl.java:87)
	at org.apache.fluss.flink.sink.writer.AppendSinkWriter.writeRow(AppendSinkWriter.java:62)
	at org.apache.fluss.flink.sink.writer.FlinkSinkWriter.write(FlinkSinkWriter.java:143)
	... 14 more
Caused by: java.lang.NullPointerException
	at org.apache.fluss.cluster.Cluster.getBucketCount(Cluster.java:177)
	at org.apache.fluss.client.write.StickyBucketAssigner.nextBucket(StickyBucketAssigner.java:76)
	at org.apache.fluss.client.write.StickyBucketAssigner.assignBucket(StickyBucketAssigner.java:50)
	at org.apache.fluss.client.write.DynamicBucketAssigner.assignBucket(DynamicBucketAssigner.java:32)
	at org.apache.fluss.client.write.StickyBucketAssigner.assignBucket(StickyBucketAssigner.java:34)
	at org.apache.fluss.client.write.WriterClient.doSend(WriterClient.java:189)
	... 19 more

```

### Solution

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-client/src/main/java/org/apache/fluss/client/write/RoundRobinBucketAssigner.java
+++ b/fluss-client/src/main/java/org/apache/fluss/client/write/RoundRobinBucketAssigner.java
@@ -31,10 +31,12 @@ import java.util.concurrent.atomic.AtomicInteger;
 @Internal
 public class RoundRobinBucketAssigner extends DynamicBucketAssigner {
     private final PhysicalTablePath physicalTablePath;
+    private final int bucketNumber;
     private final AtomicInteger counter = new AtomicInteger(new Random().nextInt());
 
-    public RoundRobinBucketAssigner(PhysicalTablePath physicalTablePath) {
+    public RoundRobinBucketAssigner(PhysicalTablePath physicalTablePath, int bucketNumber) {
         this.physicalTablePath = physicalTablePath;
+        this.bucketNumber = bucketNumber;
     }
 
     @Override
@@ -47,8 +49,7 @@ public class RoundRobinBucketAssigner extends DynamicBucketAssigner {
             return bucketsForTable.get(bucket).getBucketId();
         } else {
             // no buckets are available, give a non-available bucket.
-            return MathUtils.toPositive(nextValue)
-                    % cluster.getBucketCount(physicalTablePath.getTablePath());
+            return MathUtils.toPositive(nextValue) % bucketNumber;
         }
     }
 

--- a/fluss-client/src/main/java/org/apache/fluss/client/write/StickyBucketAssigner.java
+++ b/fluss-client/src/main/java/org/apache/fluss/client/write/StickyBucketAssigner.java
@@ -35,10 +35,12 @@ import java.util.concurrent.atomic.AtomicInteger;
 public class StickyBucketAssigner extends DynamicBucketAssigner {
 
     private final PhysicalTablePath physicalTablePath;
+    private final int bucketNumber;
     private final AtomicInteger currentBucketId;
 
-    public StickyBucketAssigner(PhysicalTablePath physicalTablePath) {
+    public StickyBucketAssigner(PhysicalTablePath physicalTablePath, int bucketNumber) {
         this.physicalTablePath = physicalTablePath;
+        this.bucketNumber = bucketNumber;
         this.currentBucketId = new AtomicInteger(-1);
     }
 
@@ -73,7 +75,7 @@ public class StickyBucketAssigner extends DynamicBucketAssigner {
                     cluster.getAvailableBucketsForPhysicalTablePath(physicalTablePath);
             if (availableBuckets.isEmpty()) {
                 int random = MathUtils.toPositive(ThreadLocalRandom.current().nextInt());
-                newBucket = random % cluster.getBucketCount(physicalTablePath.getTablePath());
+                newBucket = random % bucketNumber;
             } else if (availableBuckets.size() == 1) {
                 newBucket = availableBuckets.get(0).getBucketId();
             } else {

--- a/fluss-client/src/main/java/org/apache/fluss/client/write/WriterClient.java
+++ b/fluss-client/src/main/java/org/apache/fluss/client/write/WriterClient.java
@@ -354,9 +354,9 @@ public class WriterClient {
             ConfigOptions.NoKeyAssigner noKeyAssigner =
                     conf.get(ConfigOptions.CLIENT_WRITER_BUCKET_NO_KEY_ASSIGNER);
             if (noKeyAssigner == ROUND_ROBIN) {
-                return new RoundRobinBucketAssigner(physicalTablePath);
+                return new RoundRobinBucketAssigner(physicalTablePath, bucketNumber);
             } else if (noKeyAssigner == STICKY) {
-                return new StickyBucketAssigner(physicalTablePath);
+                return new StickyBucketAssigner(physicalTablePath, bucketNumber);
             } else {
                 throw new IllegalArgumentException(
                         "Unsupported append only row bucket assigner: " + noKeyAssigner);

--- a/fluss-client/src/test/java/org/apache/fluss/client/write/RecordAccumulatorTest.java
+++ b/fluss-client/src/test/java/org/apache/fluss/client/write/RecordAccumulatorTest.java
@@ -343,7 +343,8 @@ class RecordAccumulatorTest {
         int batchSize = 100;
         IndexedRow row = indexedRow(DATA1_ROW_TYPE, new Object[] {1, "a"});
 
-        StickyBucketAssigner bucketAssigner = new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH);
+        StickyBucketAssigner bucketAssigner =
+                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH, 3);
         RecordAccumulator accum =
                 createTestRecordAccumulator(
                         (int) Duration.ofMinutes(1).toMillis(),

--- a/fluss-client/src/test/java/org/apache/fluss/client/write/StickyStaticBucketAssignerTest.java
+++ b/fluss-client/src/test/java/org/apache/fluss/client/write/StickyStaticBucketAssignerTest.java
@@ -65,7 +65,7 @@ class StickyStaticBucketAssignerTest {
         // init cluster.
         Cluster cluster = updateCluster(Arrays.asList(bucket1, bucket2, bucket3));
         StickyBucketAssigner stickyBucketAssigner =
-                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH);
+                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH, 3);
         int bucketId = stickyBucketAssigner.assignBucket(cluster);
         assertThat(bucketId >= 0 && bucketId < 3).isTrue();
 
@@ -94,7 +94,7 @@ class StickyStaticBucketAssignerTest {
         // init cluster.
         Cluster cluster = updateCluster(Arrays.asList(bucket1, bucket2, bucket3));
         StickyBucketAssigner stickyBucketAssigner =
-                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH);
+                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH, 3);
         int bucketId = stickyBucketAssigner.assignBucket(cluster);
         for (int i = 0; i < 3; i++) {
             if (i != bucketId) {
@@ -111,7 +111,7 @@ class StickyStaticBucketAssignerTest {
         // init cluster.
         Cluster cluster = updateCluster(Collections.singletonList(bucket1));
         StickyBucketAssigner stickyBucketAssigner =
-                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH);
+                new StickyBucketAssigner(DATA1_PHYSICAL_TABLE_PATH, 3);
         int bucketId = stickyBucketAssigner.assignBucket(cluster);
 
         for (int i = 0; i < 100; i++) {
@@ -136,7 +136,7 @@ class StickyStaticBucketAssignerTest {
         Cluster cluster = updateCluster(allBuckets);
 
         // Assure we never choose bucket 1 for tp1 because it is unavailable.
-        StickyBucketAssigner stickyBucketAssigner = new StickyBucketAssigner(tp1);
+        StickyBucketAssigner stickyBucketAssigner = new StickyBucketAssigner(tp1, 3);
         int bucketForTp1 = stickyBucketAssigner.assignBucket(cluster);
         assertThat(bucketForTp1).isNotEqualTo(1);
         for (int i = 0; i < 100; i++) {
@@ -145,7 +145,7 @@ class StickyStaticBucketAssignerTest {
         }
 
         // Assure we always choose bucket 1 for tp2.
-        stickyBucketAssigner = new StickyBucketAssigner(tp2);
+        stickyBucketAssigner = new StickyBucketAssigner(tp2, 3);
         int bucketForTp2 = stickyBucketAssigner.assignBucket(cluster);
         assertThat(bucketForTp2).isEqualTo(1);
         for (int i = 0; i < 100; i++) {
@@ -154,7 +154,7 @@ class StickyStaticBucketAssignerTest {
         }
 
         // Assure that we can still choose one bucket even if there are no available buckets.
-        stickyBucketAssigner = new StickyBucketAssigner(tp3);
+        stickyBucketAssigner = new StickyBucketAssigner(tp3, 3);
         int bucketForTp3 = stickyBucketAssigner.assignBucket(cluster);
         assertThat(bucketForTp3).isIn(0, 1, 2);
         stickyBucketAssigner.onNewBatch(cluster, bucketForTp3);
@@ -165,7 +165,7 @@ class StickyStaticBucketAssignerTest {
     void testMultiThreadToCallOnNewBatch() {
         Cluster cluster = updateCluster(Arrays.asList(bucket1, bucket2, bucket3));
         StickyBucketAssigner stickyBucketAssigner =
-                new StickyBucketAssigner(PhysicalTablePath.of(DATA1_TABLE_PATH));
+                new StickyBucketAssigner(PhysicalTablePath.of(DATA1_TABLE_PATH), 3);
         int bucketId = stickyBucketAssigner.assignBucket(cluster);
         Queue<Integer> bucketIds = new ConcurrentLinkedQueue<>();
         Thread[] threads = new Thread[100];

--- a/fluss-common/src/main/java/org/apache/fluss/cluster/Cluster.java
+++ b/fluss-common/src/main/java/org/apache/fluss/cluster/Cluster.java
@@ -173,10 +173,6 @@ public final class Cluster {
                                                 + " in cluster"));
     }
 
-    public int getBucketCount(TablePath tablePath) {
-        return tableInfoByPath.get(tablePath).getNumBuckets();
-    }
-
     /** Get the bucket location for this table-bucket. */
     public Optional<BucketLocation> getBucketLocation(TableBucket tableBucket) {
         return Optional.ofNullable(availableLocationByBucket.get(tableBucket));

